### PR TITLE
mon/MonmapMonitor: do not propose on error in prepare_update

### PR DESCRIPTION
**Pull request by batrick**
Also: correct a serious protocol error where the monitor would reply to a command before the proposal is committed.

Fixes: https://tracker.ceph.com/issues/58974

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by ajarr — COMMENTED** [review]
I think the commit could be broken down into three separate commits that solve three separate issues.
* MonMapMonitor: do not propose changes on error in prepare_update/prepare_command

* MonMapMonitor: commit proposed map changes initiated by a `ceph mon` command before replying
(This seems like a more important fix than the one above. It changes the monitor's handling of `ceph mon` commands that has existed since https://github.com/ceph/ceph/pull/6854)

* MonMapMonitor: do not propose changes when `mon enable msgr-2` command is called on monitors that already have msgr2 enabled

File: src/mon/MonmapMonitor.cc
Comment (batrick) [reply src/mon/MonmapMonitor.cc:510]: Good catch, thanks!

**Review by ajarr — APPROVED** [review]
No longer need "Also: correct a serious protocol error where the monitor would reply to
a command before the proposal is committed" in the commit message of https://github.com/ceph/ceph/pull/50503/commits/54406341071aec299dd2b9a0b37d8d6c5d641bd6" ?

**Comment by ajarr** [comment]
> > ```
> > * MonMapMonitor: do not propose changes when `mon enable msgr-2` command is called on monitors that already have msgr2 enabled
> > ```
> 
> I restructured the code but I do not think it proposed when `mon enable msgr-2` was a no-op. So I don't think this separate commit makes sense?

Yes, makes sense.

**Comment by batrick** [comment]
> No longer need "Also: correct a serious protocol error where the monitor would reply to a command before the proposal is committed" in the commit message of [5440634](https://github.com/ceph/ceph/commit/54406341071aec299dd2b9a0b37d8d6c5d641bd6)" ?

good point. Fixed

**Comment by batrick** [comment]
@yuriw please include in the next RADOS run.

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by ljflores** [comment]
Hey @batrick, I'm seeing some failures that look related:

Slow ops from a mon command that was changed in this PR
/a/yuriw-2023-07-03_15:28:45-rados-wip-yuri-testing-2023-06-23-0831-distro-default-smithi/7324842
```
2023-07-03T20:55:47.552 INFO:tasks.ceph.mon.b.smithi123.stderr:2023-07-03T20:55:47.550+0000 7f838f917700 -1 mon.b@1(electing) e5685 get_health_metrics reporting 1 slow ops, oldest is mon_command({"prefix": "mon set election_strategy", "strategy": "disallow"} v 0)
2023-07-03T20:55:52.552 INFO:tasks.ceph.mon.b.smithi123.stderr:2023-07-03T20:55:52.550+0000 7f838f917700 -1 mon.b@1(electing) e5686 get_health_metrics reporting 1 slow ops, oldest is mon_command({"prefix": "mon set election_strategy", "strategy": "disallow"} v 0)
2023-07-03T20:55:57.552 INFO:tasks.ceph.mon.b.smithi123.stderr:2023-07-03T20:55:57.551+0000 7f838f917700 -1 mon.b@1(electing) e5687 get_health_metrics reporting 1 slow ops, oldest is mon_command({"prefix": "mon set election_strategy", "strategy": "disallow"} v 0)
2023-07-03T20:56:02.553 INFO:tasks.ceph.mon.b.smithi123.stderr:2023-07-03T20:56:02.550+0000 7f838f917700 -1 mon.b@1(electing) e5689 get_health_metrics reporting 1 slow ops, oldest is mon_command({"prefix": "mon set election_strategy", "strategy": "disallow"} v 0)
2023-07-03T20:56:07.553 INFO:tasks.ceph.mon.b.smithi123.stderr:2023-07-03T20:56:07.551+0000 7f838f917700 -1 mon.b@1(electing) e5690 get_health_metrics reporting 1 slow ops, oldest is mon_command({"prefix": "mon set election_strategy", "strategy": "disallow"} v 0)
2023-07-03T20:56:12.266 INFO:tasks.workunit.client.0.smithi123.stderr://home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1: test_mon_mon:  rm -fr /tmp/cephtool.svX
```

Here was the test link:
http://pulpito.front.sepia.ceph.com/?branch=wip-yuri-testing-2023-06-23-0831

And here are a few more jobs that were affected:
http://pulpito.front.sepia.ceph.com/yuriw-2023-07-03_15:28:45-rados-wip-yuri-testing-2023-06-23-0831-distro-default-smithi/7324842/
http://pulpito.front.sepia.ceph.com/yuriw-2023-07-03_15:28:45-rados-wip-yuri-testing-2023-06-23-0831-distro-default-smithi/7324846/
http://pulpito.front.sepia.ceph.com/yuriw-2023-07-03_15:28:45-rados-wip-yuri-testing-2023-06-23-0831-distro-default-smithi/7324850/
http://pulpito.front.sepia.ceph.com/yuriw-2023-07-03_15:28:45-rados-wip-yuri-testing-2023-06-23-0831-distro-default-smithi/7324858/

**Comment by batrick** [comment]
@ljflores this is ready for QA again but please don't QA with Ilya's PR again. :laughing:

**Comment by batrick** [comment]
So another added benefit I didn't plan for is that the commits mentioned in https://github.com/ceph/ceph/pull/50503#issuecomment-1634815314 make the replies from the monitors significantly faster. For example, on a humble vstart cluster:

```
2023-07-13T15:37:17.196-0400 7fe7e26ea700  0 mon.a@0(leader) e1 handle_command mon_command({"prefix": "osd pool create", "pool": "cephfs.b.meta"} v 0) v1
...
2023-07-13T15:37:17.278-0400 7fe7dfee5700 10 mon.a@0(leader) e1 refresh_from_paxos
2023-07-13T15:37:17.278-0400 7fe7dfee5700  0 log_channel(audit) log [INF] : from='mgr.4127 ' entity='mgr.x' cmd='[{"prefix": "osd pool create", "pool": "cephfs.b.meta"}]': finished
2023-07-13T15:37:17.278-0400 7fe7dfee5700  1 -- [v2:192.168.230.102:40693/0,v1:192.168.230.102:40694/0] --> [v2:192.168.230.102:40693/0,v1:192.168.230.102:40694/0] -- log(1 entries from seq 268 at 2023-07-13T15:37:17.279871-0400) v1 -- 0x5597f6e288c0 con 0x5597f1530c00
2023-07-13T15:37:17.278-0400 7fe7dfee5700  2 mon.a@0(leader) e1 send_reply 0x5597f6e1f320 0x5597f69baea0 mon_command_ack([{"prefix": "osd pool create", "pool": "cephfs.b.meta"}]=0 pool 'cephfs.b.meta' created v83) v1
...
2023-07-13T15:37:17.286-0400 7fe7dfee5700 10 mon.a@0(leader).osd e83 create_pending e 84
2023-07-13T15:37:17.286-0400 7fe7dfee5700 10 mon.a@0(leader).osd e83 update_logger
```

RTT is 82ms instead of what would normally be ~90ms. That's an ~9% improvement. Note replies are currently only sent after pending is created.

It can be hard to benchmark this because mon command routing can add latency. I have a test which creates a single RADOS session but I haven't yet figured out how to force it to connect to the leader for the benchmark.

**Comment by batrick** [comment]
jenkins test api

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcVbn2qwgo1793-wip-yuri-testing-2023-07-14-1641-old-wip-yuri-testing-2023-07-12-1332-old-wip-yuri-testing-2023-07-12-1140-old-wip-yuri-testing

Let me know if it needs testing again.

**Comment by batrick** [comment]
```
main branch:
min/max/mean/stddev: 0.015813/0.613919/0.031615/0.024087
min/max/mean/stddev: 0.015737/0.255008/0.031492/0.017930
min/max/mean/stddev: 0.014242/0.205763/0.031969/0.018022
min/max/mean/stddev: 0.014172/0.270256/0.032070/0.021079
min/max/mean/stddev: 0.017767/0.471187/0.032751/0.025317

upstream/pull/50503/head:
min/max/mean/stddev: 0.010476/0.158475/0.026324/0.013662
min/max/mean/stddev: 0.016866/0.099403/0.027938/0.010508
min/max/mean/stddev: 0.014013/0.127512/0.026847/0.010340
min/max/mean/stddev: 0.013098/0.172725/0.028979/0.012998
min/max/mean/stddev: 0.016934/0.292218/0.029252/0.014904
```

About a ~20% improvement in latency!

**Comment by batrick** [comment]
@ljflores I think this is ready for QA but another review by a core team member would be appropriate.

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
rebased + commit message changes

**Comment by batrick** [comment]
jenkins test api

**Comment by ljflores** [comment]
jenkins test api

**Review by ljflores — APPROVED** [review]
Looks good, I like the performance analysis you added!

**Review by rzarzynski — COMMENTED** [review]
LGTM.

**Comment by rzarzynski** [comment]
Adding `needs-qa`. When it comes to backporting, I would prefer to bake these changes in `main` for a while.

**Comment by rishabh-d-dave** [comment]
@batrick If you haven't started testing this already, I can put it through testing in the run I am starting today/tomorrow.

**Comment by batrick** [comment]
Nothing out of the ordinary from this PR: https://tracker.ceph.com/projects/cephfs/wiki/Main#2023-Sep-12

I think we can merge this unless @ljflores @rzarzynski you would like more `rados` suite testing.

**Comment by ljflores** [comment]
Apologies @batrick, this was tested and rados-approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcpMEWaauy1825-wip-yuri3-testing-2023-08-15-0955

I missed adding a comment here.

**Pull request merged**
